nixos/modules/services/torrent/transmission.nix

```diff
@@ -104,47 +99,79 @@ in
   };
 
   config = mkIf cfg.enable {
-    systemd.tmpfiles.rules = [
-      "d '${homeDir}' 0770 '${cfg.user}' '${cfg.group}' - -"
-      "d '${settingsDir}' 0700 '${cfg.user}' '${cfg.group}' - -"
-      "d '${fullSettings.download-dir}' '${downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
-      "d '${fullSettings.incomplete-dir}' '${downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+    systemd.tmpfiles.rules =
+      optionals (! hasPrefix "/var/lib/" cfg.home) [
+        "d '${cfg.home}' 755 '${cfg.user}' '${cfg.group}' - -"
+        # Avoid "Unsafe path transition" by specifying parent directory .config
+        "d '${cfg.home}/.config' 755 '${cfg.user}' '${cfg.group}' - -"
+        "d '${cfg.home}/${settingsDir}' 700 '${cfg.user}' '${cfg.group}' - -"
+      ] ++
+      optionals (cfg.downloadDirPermissions != null) (
+        [ "d '${cfg.settings.download-dir}' '${cfg.downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+        ] ++ optional cfg.settings.incomplete-dir-enabled
+          "d '${cfg.settings.incomplete-dir}' '${cfg.downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+      );
+
+    assertions = [
+      { assertion = builtins.match "^/.*" cfg.home != null;
+        message = "`services.transmission.home' must be an absolute path.";
+      }
     ];
 
     systemd.services.transmission = {
       description = "Transmission BitTorrent Service";
       after = [ "network.target" ] ++ optional apparmor "apparmor.service";
       requires = mkIf apparmor [ "apparmor.service" ];
       wantedBy = [ "multi-user.target" ];
-
-      # 1) Only the "transmission" user and group have access to torrents.
-      # 2) Optionally update/force specific fields into the configuration file.
-      serviceConfig.ExecStartPre = preStart;
-      serviceConfig.ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f --port ${toString config.services.transmission.port} --config-dir ${settingsDir}";
-      serviceConfig.ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
-      serviceConfig.User = cfg.user;
-      serviceConfig.Group = cfg.group;
-      # NOTE: transmission has an internal umask that also must be set (in settings.json)
-      serviceConfig.UMask = "0002";
+      preStart = ''
+        cp -f ${settingsFile} '${cfg.home}/${settingsDir}/settings.json'
```

Comment:
It's been forever since I looked at this module. Can you refresh my memory on why we can't directly reference this file from the nix store?

Reply:
@ju1m I'd like to suggest a temporary workaround for:

```
TODO: switch to configGen.json once RFC0042 is implemented
```

In order to avoid having passwords in my store, I use on my personal prestart script:

```
          ${pkgs.jq.bin}/bin/jq -s add /var/lib/secrets/transmission.json ${settingsFile} > ${settingsDir}/settings.json
```

Where `/var/lib/secrets/transmission.json` is where I keep the private part of the settings in json format. Perhaps we could add a `credentialsFile` settings that will point to such a json file and it will be used to add the private credentials as in my `jq` command?

Also, a bit related to @aanderse's question, I think this:

```
settingsFile = pkgs.writeText "settings.json" (builtins.toJSON cfg.settings);
```

Is not needed, and we could craft that `jq` command to include the configured settings via `builtins.toJSON` directly.

Reply:
@aanderse if I symlink ˋsettings.jsonˋ to the Nix store, stopping Transmission emits this message: "Couldn't save temporary file "/nix/store/yvm6g083f25kx0k58kgsb040qf8k3ll4-settings.json.tmp.Wl4otP": Read-only file system", and thus does not save the changes done through the RPC clients, but since ˋsettings.jsonˋ is reset in the ˋpreStartˋ anyway even with a copy, this is the same than a symlink AFAIU. But @doronbehar's proposal would produce a file outside the Nix store anyway, I'll add it tomorrow.

Reply:
Transmission also overwrites `settings.json` at startup, hashing `rpc-password`.

Reply:
This is resolved.